text/0000-guaranteed-slice-repr.md

```diff
@@ -0,0 +1,205 @@
+  is the common representation. This is even observable using `is_layout_compatible`
+  [on MSVC](https://godbolt.org/z/Y8ardrshY), though not
+  [on GCC](https://godbolt.org/z/s4v4xehnG) nor
+  [on Clang](https://godbolt.org/z/qsd1K5oGq). Future changes to guarantee a
+  different layout in the C++ standard (unlikely due to MSVC ABI stabilitiy
+  requirements) could preclude matching the layout with `&[T]`.
+
+* Unlike Rust, `std::span` allows the `data` pointer to be `nullptr`. One
+  possibile workaround for this would be to guarantee that `Option<&[T]>` uses
+  `data: std::ptr::null(), len: 0` to represent the `None` case, making
+  `std::span<T>` equivalent to `Option<&[T]>` for non-zero-sized types.
+
+  Note that this is not currently the case. The compiler currenty represents
+  `None::<&[u8]>` as `data: std::ptr::null(), len: uninit` (though this is
+  not guaranteed).
+
+* Rust uses a dangling pointer in the representation of zero-length slices.
+  It's unclear whether C++ guarantees that a dangling pointer will remain
+  unchanged when passed through `std::span`. However, it does support
+  dangling pointers during regular construction via the use of
+  [`std::to_address`](https://en.cppreference.com/w/cpp/container/span/span)
+  in the iterator constructors.
+
+Note that C++ also does not support zero-sized types, so there is no naive way
+to represent types like `std::span<SomeZeroSizedRustType>`.
+
+## Flexibility
+
+Additionally, guaranteeing layout of Rust-native types limits the compiler's and
+standard library's ability to change and take advantage of new optimization
+opportunities.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* We could avoid committing to a particular representation for slices.
+
+* We could try to guarantee layout compatibility with a particular target's
+  `std::span` representation, though without standardization this may be
+  impossible. Multiple different C++ stdlib implementations may be used on
+  the same platform and could potentially have different span representations.
+  In practice, current span representations also use ptr+len pairs.
+
+* We could avoid storing a data pointer for zero-sized types. This would result
+  in a more compact representation but would mean that the representation of
+  `&[T]` is dependent on the type of `T`. Additionally, this would break
+  existing code which depends on storing data in the pointer of ZST slices.
+
+  This would break popular crates such as [bitvec](https://docs.rs/crate/bitvec/1.0.1/source/doc/ptr/BitSpan.md)
+  (55 million downloads) and would result in strange behavior such as
+  `std::ptr::slice_from_raw_parts(ptr, len).as_ptr()` returning a different
+  pointer from the one that was passed in.
+
+  Types like `*const ()` / `&()` are widely used to pass around pointers today.
+  We cannot make them zero-sized, and it would be surprising to make a
+  different choice for `&[()]`.
```

Comment:
Maybe we should have warned against this and made everyone use `*const u8` instead. Seems impossible to make a change like this now though.